### PR TITLE
V3alpha no tuple

### PR DESCRIPTION
**Pull request by facontidavide**
Apply change discussed in #303

### CODE REVIEW

File: examples/graph/main.cpp [hunk 1]
```diff
@@ -4,8 +4,8 @@
 #include <QtNodes/StyleCollection>
 
 #include <QtWidgets/QApplication>
-#include <QtGui/QAction>
-#include <QtGui/QScreen>
```
Comment (facontidavide) [inline examples/graph/main.cpp:8]: I was not able to compile without his change

File: include/QtNodes/internal/ConnectionGraphicsObject.hpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ class ConnectionGraphicsObject : public QGraphicsObject
   BasicGraphicsScene *
   nodeScene() const;
 
-  ConnectionId
```
Comment (facontidavide) [inline include/QtNodes/internal/ConnectionGraphicsObject.hpp:43]: No need to make a copy every time

File: examples/graph/main.cpp (hunk 1 above)
Comment (paceholder) [reply examples/graph/main.cpp:8]: That means you use something prior to 5.15

File: examples/graph/CustomGraphModel.cpp [hunk 3]
```diff
@@ -40,17 +40,13 @@ allConnectionIds(NodeId const nodeId) const
...
+                       target.first, target.second});
       }
       else
       {
-        result.insert(std::make_tuple(target.first,
-                                      target.second,
-                                      nodeId,
-                                      portIndex));
+        result.insert({target.first, target.second,
+                       nodeId, portIndex});
```
Comment (paceholder) [inline examples/graph/CustomGraphModel.cpp:49]: Maybe `result.insert(ConnectionId{nodeId, portIndex, target.first, target.second});`

File: include/QtNodes/internal/Definitions.hpp [hunk 4]
```diff
@@ -98,10 +98,36 @@ static constexpr NodeId InvalidNodeId =
...
+         a.inPortIndex == b.inPortIndex;
+}
+
+inline bool operator!=(const ConnectionId& a, const ConnectionId& b)
+{
+  return !(a == b);
+}
+
+
+inline void SwapPorts(ConnectionId& id)
```
Comment (paceholder) [inline include/QtNodes/internal/Definitions.hpp:126]: `swapPorts`

File: src/BasicGraphicsScene.cpp [hunk 5]
```diff
@@ -203,8 +203,7 @@ traverseGraphAndPopulateGraphicsObjects()
           fifo.push(cn.first);
           allNodeIds.erase(cn.first);
 
-          auto connectionId = std::make_tuple(nodeId, index, cn.first, cn.second);
-
+          ConnectionId connectionId ={nodeId, index, cn.first, cn.second};
```
Comment (paceholder) [inline src/BasicGraphicsScene.cpp:206]: `ConnectionId connectionId{nodeId, index, cn.first, cn.second};`

File: src/ConnectionGraphicsObject.cpp [hunk 6]
```diff
@@ -106,7 +106,7 @@ nodeScene() const
 }
 
 
-ConnectionId
+const ConnectionId &
```
Comment (paceholder) [inline src/ConnectionGraphicsObject.cpp:109]: `ConnectionId const &`
https://mariusbancila.ro/blog/2018/11/23/join-the-east-const-revolution/

File: src/ConnectionState.cpp [hunk 7]
```diff
@@ -40,9 +40,9 @@ bool
 ConnectionState::
 requiresPort() const
 {
-  ConnectionId id = _cgo.connectionId();
-  return std::get<0>(id) == InvalidNodeId ||
-         std::get<2>(id) == InvalidNodeId;
+  const ConnectionId& id = _cgo.connectionId();
```
Comment (paceholder) [inline src/ConnectionState.cpp:43]: `ConnectionId const&`

File: src/NodeGraphicsObject.cpp [hunk 8]
```diff
@@ -179,10 +179,9 @@ moveConnections() const
         for (auto& cn: connectedNodes)
         {
           // out node id, out port index, in node id, in port index.
-          ConnectionId connectionId =
-            (portType == PortType::In) ?
-            std::make_tuple(cn.first, cn.second, _nodeId, portIndex) :
-            std::make_tuple(_nodeId, portIndex, cn.first, cn.second);
+          ConnectionId connectionId = {cn.first, cn.second, _nodeId, portIndex};
```
Comment (paceholder) [inline src/NodeGraphicsObject.cpp:182]: `ConnectionId connectionId{cn.first, cn.second, _nodeId, portIndex};`

File: src/NodeGraphicsObject.cpp [hunk 9]
```diff
@@ -266,8 +265,7 @@ mousePressEvent(QGraphicsSceneMouseEvent* event)
         auto const& cn = *connectedNodes.begin();
 
         // Need "reversed" connectin id if enabled for both port types.
-        ConnectionId connectionId =
-          std::make_tuple(cn.first, cn.second, _nodeId, portIndex);
+        ConnectionId connectionId = {cn.first, cn.second, _nodeId, portIndex};
```
Comment (paceholder) [inline src/NodeGraphicsObject.cpp:268]: ConnectionId connectionId{cn.first, cn.second, _nodeId, portIndex};

File: src/NodeGraphicsObject.cpp [hunk 10]
```diff
@@ -293,8 +291,7 @@ mousePressEvent(QGraphicsSceneMouseEvent* event)
           {
             for (auto& cn : connectedNodes)
             {
-              ConnectionId connectionId =
-                std::make_tuple(_nodeId, portIndex, cn.first, cn.second);
+              ConnectionId connectionId = {_nodeId, portIndex, cn.first, cn.second};
```
Comment (paceholder) [inline src/NodeGraphicsObject.cpp:294]: `ConnectionId connectionId{_nodeId, portIndex, cn.first, cn.second};`

File: include/QtNodes/internal/Definitions.hpp (hunk 4 above)
Comment (facontidavide) [reply include/QtNodes/internal/Definitions.hpp:126]: sounds good to me, but should be a verb: **invertConnection**

File: src/BasicGraphicsScene.cpp (hunk 5 above)
Comment (facontidavide) [reply src/BasicGraphicsScene.cpp:206]: same.... can change it

File: src/ConnectionGraphicsObject.cpp (hunk 6 above)
Comment (facontidavide) [reply src/ConnectionGraphicsObject.cpp:109]: oh, you are one of those ;)

**Comment by facontidavide** [comment]
by the way, may I suggest to add a clang-format file, to make future PR easier?

**Comment by paceholder** [comment]
A clang-format config would be perfect. Even if we might break some nuances of the current library formatting.
I did not use this tool personally yet, that's why I postpone working on a config again and again.

**Pull request merged**